WalletWasabi/Services/SingleInstanceChecker.cs

```diff
@@ -1,95 +1,189 @@
+				using CancellationTokenSource cts = CancellationTokenSource.CreateLinkedTokenSource(DisposeCts.Token, timeoutCts.Token);
+
+				// Make sure non cancel-able operations are aborted.
+				using var _ = cts.Token.Register(() => client.Dispose());
+
+				await client.ConnectAsync(IPAddress.Loopback, Port, cts.Token).ConfigureAwait(false);
+
+				await using NetworkStream networkStream = client.GetStream();
+				networkStream.WriteTimeout = (int)ClientTimeOut.TotalMilliseconds * 2;
+				await using var writer = new StreamWriter(networkStream, Encoding.UTF8);
+				await writer.WriteAsync(new StringBuilder(WasabiMagicString), cts.Token).ConfigureAwait(false);
+
+				// I was able to signal to the other instance successfully so just continue.
 			}
-			catch (IOException ex)
+			catch (Exception ex)
 			{
-				throw new InvalidOperationException($"Wasabi is already running on {Network}!", ex);
+				// Do not log anything here as the first instance is writing the Log at this time.
+				throw new InvalidOperationException($"Wasabi is already running, but cannot be signaled, reason: '{ex}'");
 			}
+
+			throw new OperationCanceledException($"Wasabi is already running, signaled the first instance.");
 		}
 
-		/// <summary>
-		/// <list type="bullet">
-		/// <item>Unmanaged resources need to be released regardless of the value of the <paramref name="disposing"/> parameter.</item>
-		/// <item>Managed resources need to be released if the value of <paramref name="disposing"/> is <c>true</c>.</item>
-		/// </list>
-		/// </summary>
-		/// <param name="disposing">
-		/// Indicates whether the method call comes from a <see cref="Dispose()"/> method
-		/// (its value is <c>true</c>) or from a finalizer (its value is <c>false</c>).
-		/// </param>
-		protected virtual void Dispose(bool disposing)
+		private static int NetworkToPort(Network network) => network switch
+		{
+			_ when network == Network.Main => 37129,
+			_ when network == Network.TestNet => 37130,
+			_ when network == Network.RegTest => 37131,
+			_ => throw new Exception($"Network {network} is unknown")
+		};
+
+		protected override async Task ExecuteAsync(CancellationToken stoppingToken)
 		{
-			if (!_disposedValue)
+			var task = TaskStartTcpListener;
+			if (task is null)
 			{
-				if (disposing)
+				throw new InvalidOperationException("This is impossible!");
+			}
+
+			TcpListener? listener = null;
+			try
+			{
+				listener = new(IPAddress.Loopback, Port);
+
+				// This can throw an exception if the port is already open.
+				listener.Start(0);
+
+				// Indicate that the Listener is created successfully.
+				task.TrySetResult();
+
+				// Stop listener here to ensure thread-safety.
+				using var _ = stoppingToken.Register(() => listener.Stop());
+
+				while (!stoppingToken.IsCancellationRequested)
 				{
-					SingleApplicationLockHolder?.Dispose();
+					// In case of cancellation, listener.Stop will cause AcceptTcpClientAsync to throw, thus canceling it.
```

Comment:
I'm missing how `listener.AcceptTcpClientAsync` can get canceled. Could you explain it? I think it will stuck waiting for the client if there is no one.

There is https://github.com/dotnet/runtime/issues/43935 issue which tracks adding the cancellation token to this method. Maybe we should add it here so that once we are on .NET 6 (November 2021) or later we can change it.

Reply:
I would guess that `.WithAwaitCancellationAsync(som-token)` should be here too.

Reply:
It was canceled by listener.Dispose with the CancelToken.Register(). I removed all of those and use WithAwaitCancellation.